### PR TITLE
[feat][txn]Implement transactional consumer/producer API

### PR DESCRIPTION
**Pull request by liangyepianzhou**
Master Issue:https://github.com/apache/pulsar-client-go/issues/932
### Motivation
Implement transaction coordinator client.
### Modifications
1. Implement transaction coordinator
2. implement transactionImpl
3. Implement transaction in producer and consumer API

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)

### Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / GoDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

**Comment by shibd** [comment]
Is this the last PR about transactions? It's better to add some sample code to examples.

**Comment by liangyepianzhou** [comment]
>Is this the last PR about transactions? It's better to add some sample code to examples.

Yes, this is the last PR concerning transactions. In the tests I've added, I've only used the API to perform the tests, which can serve as an initial example. Adding more detailed examples to the documentation might be more appropriate?

**Comment by shibd** [comment]
> Adding more detailed examples to the documentation might be more appropriate?

Yes, the documentation can be described step by step, and complete examples are placed in the examples folder. We can push in a later PR.

File: pulsar/client_impl.go [hunk 1]
```diff
@@ -195,6 +195,11 @@ func newClient(options ClientOptions) (Client, error) {
 	return c, nil
 }
 
+func (c *client) NewTransaction(timeout time.Duration) (Transaction, error) {
+	id, err := c.tcClient.newTransaction(timeout)
```
Comment (BewareMyPower) [inline pulsar/client_impl.go:199]: We should check `err` here:

```go
        if id, err := c.tcClient.newTransaction(timeout); err == nil {
                return newTransaction(*id, c.tcClient, timeout), nil
        } else {
                return nil, err
        }
```

File: pulsar/consumer_partition.go [hunk 2]
```diff
@@ -417,6 +417,93 @@ func (pc *partitionConsumer) Unsubscribe() error {
 	return req.err
 }
 
+func (pc *partitionConsumer) AckIDWithTxn(msgID MessageID, txn Transaction) error {
```
Comment (BewareMyPower) [inline pulsar/consumer_partition.go:420]: Could you reduce the duplicated code with `ackID`? I see much code is just copied from `ackID`.

**Comment by Anonymitaet** [comment]
Hi @liangyepianzhou thanks for introducing this great feature!

========================================

As we discussed just now, for the doc side:

### ✅ Pulsar feature matrix

I've ticked the three boxes (client, producer, consumer) of ["Transactions" for Go](https://docs.google.com/spreadsheets/d/1YHYTkIXR8-Ql103u-IMI18TXLlGStK8uJjDsOOA0T20/edit#gid=1784579914&range=19:19).

### ✅ Pulsar client docs

1. Could you please provide the tech inputs as below?
  
    Add Go code example to [Transactions > Get started ](https://pulsar.apache.org/docs/next/txn-use/).

image

2. This feature is available from Pulsar 2.9.5, can we add the doc to 4 doc versions (NEXT, 2.11.x, 2.10.x, 2.9.x) or just indicate what version it supports?

    - For 2.9.x: we can add a tip to the 2.9.x doc and indicate that this feature is available from (or in) Pulsar 2.9.5.
    - For 2.11.x and 2.10.x: please see the questions below.
image

BTW,
- Original doc file of NEXT is [here](https://github.com/apache/pulsar-site/edit/main/docs).
- Original doc files of all versioned docs are [here](https://github.com/apache/pulsar-site/tree/main/versioned_docs).

========================================

Feel free to reach out to me if you have questions. TIA!

**Comment by liangyepianzhou** [comment]
Could you please retake a look? @BewareMyPower @shibd

**Comment by shibd** [comment]
> Could you please retake a look? @BewareMyPower @shibd

Sorry, I'll take a look at it today.

**Review by BewareMyPower — APPROVED** [review]

File: pulsar/transaction_test.go [hunk 3]
```diff
@@ -217,3 +218,109 @@ func createTcClient(t *testing.T) (*transactionCoordinatorClient, *client) {
...
+// 3. Produce 10 messages with a transaction and 10 messages without a transaction.
+// - Expectation: The consumer should be able to receive the 10 messages sent without a transaction,
+// but not the 10 messages sent with the transaction.
+// 4. Commit the transaction and receive the remaining 10 messages.
+// - Expectation: The consumer should be able to receive the 10 messages sent with the transaction.
+// 5. Clean up: Close the consumer and producer instances.
+//
+// The test ensures that the consumer can only receive messages sent with a transaction after it is committed,
+// and that it can always receive messages sent without a transaction.
+func TestConsumeAndProduceWithTxn(t *testing.T) {
```
Comment (shibd) [inline pulsar/transaction_test.go:236]: This unit test cover flow scenarios:

1. Producer sends some message to the topic, before committing txn. All messages are not visible to the consumer.
2. When consumer receive message, use `ackWithTxn` to ack the message together.

I think we need more unite tests to cover the transaction feature:

1. SourceTopic -> **consumer.ack and producer.send** -> SinkTopic. We need to guarantee `ack source topic message` and `send message to sink topic` on a transaction. 
2. Transaction abort?

**Review by shibd — COMMENTED** [review]
Overall LGTM, left some small comments.

File: pulsar/transaction_test.go (hunk 3 above)
Comment (liangyepianzhou) [reply pulsar/transaction_test.go:236]: You are correct that more unit tests are needed to cover various scenarios when working with transactions.

**Review by shibd — APPROVED** [review]

**Pull request merged**